Scripts/World/world_state.gd

```diff
@@ -0,0 +1,716 @@
+		result.player_states[i] = WorldPlayerState.new()
+		var player = result.player_states[i]
+		var slot = slots[i]
+		player.faction = slot.faction
+		player.goods = CFG.get_start_goods()
+
+	for x in map.grid_width:
+		for y in map.grid_height:
+			var coord := Vector2i(x, y)
+			var hex : WorldHex = WorldHex.new()
+			hex.data_tile = map.grid_data[x][y]
+			var place_ser : Dictionary = {}
+			var type : String = hex.data_tile.type
+			if ser:
+				place_ser = ser.place_hexes[coord]
+				type = place_ser["type"]
+			hex.init_place(type, coord, place_ser)
+			# TODO somehow get this inside of
+			# creation
+			result.grid.set_hex(coord, hex)
+			if not ser and hex.place:
+				var army_preset = hex.place.get_army_at_start()
+				if army_preset:
+					result.spawn_army_from_preset(army_preset, coord, \
+						hex.place.controller_index)
+			if ser and coord in ser.army_hexes:
+				var loaded : Dictionary = ser.army_hexes[coord]
+
+
+	result.synchronize_players_with_their_places()
+
+
+	#if world_state:
+		#for i in players.size():
+			#var ser = world_state.players[i]
+			#var player : Player = players[i]
+			## swap cities if capital is not first
+			#if player.capital_city.coord != ser.capital_city:
+				#pass # TODO -- previous version was probably buggy
+			#player.goods = Goods.from_array(ser.goods)
+			#for identifier in ser.dead_heroes:
+				#var hero := Hero.from_network_serializable(identifier, player)
+				#player.dead_heroes.append(hero)
+		#current_player = players[world_state.current_player]
+
+	result.current_player_index = 0
+
+	return result
+
+
+## fills up players' fields like `cities` and `outposts` after place grid is
+## filled with new map or loaded state
+func synchronize_players_with_their_places() -> void:
+	for player in player_states:
+		player.cities = []
+		player.outposts = []
+	for x in grid.width:
+		for y in grid.height:
+			var coord := Vector2i(x, y)
+			var city = get_city_at(coord)
+			var outpost = get_place_at(coord) as Outpost
+			if city:
+				var player = get_player(city.controller_index)
+				if player:
+					player.cities.append(city)
+			if outpost:
+				var player = get_player(outpost.controller_index)
+				if player:
+					player.outposts.append(outpost)
+
+
+
+#endregion
+
+
+func get_current_player() -> WorldPlayerState:
+	if current_player_index in range(player_states.size()):
+		return player_states[current_player_index]
+	return null
+
+
+func get_player_index(player : WorldPlayerState) -> int:
+	for i in range(player_states.size()):
+		if player == player_states[i]:
+			return i
+	return -1
+
+
+func get_player(index : int) -> WorldPlayerState:
+	return get_player_by_index(index)
+
+
+func get_player_by_index(index : int) -> WorldPlayerState:
+	if index < 0:
+		return null
+	if index >= player_states.size():
+		return null
+	return player_states[index]
+
+
+func check_move_allowed(world_move_info : WorldMoveInfo) -> String:
+	if world_move_info.move_type == WorldMoveInfo.TYPE_TRAVEL:
```

Comment:
with this big of a switch WorldMoveInfo should have subclasses to handle this behaviour

Reply:
yes i thought about this, but do not have an idea about this -- maybe next PR

Reply:
also WorldMoveInfo is move like `struct` than `class` type, so adding inheritance to this will make it overcomplicated -- we need to find some other way of dividing it

Reply:
there are too many types, its easier to manage and debug with small classes